pandas/core/internals/blocks.py

```diff
@@ -801,10 +801,20 @@ def _replace_list(
 
         rb = [self if inplace else self.copy()]
         for i, (src, dest) in enumerate(pairs):
+            convert = i == src_len  # only convert once at the end
             new_rb: List[Block] = []
-            for blk in rb:
-                m = masks[i]
-                convert = i == src_len  # only convert once at the end
+
+            # GH-39338: _replace_coerce can split a block into
+            # single-column blocks, so track the index so we know
+            # where to index into the mask
```

Comment:
may be hard to backport as this code has changed a lot (though possible *this* particular section has not).